src/distilabel/pipeline/templates/instruction.py

```diff
@@ -0,0 +1,136 @@
+
+from distilabel.distiset import Distiset
+from distilabel.llms.base import LLM
+from distilabel.llms.huggingface import InferenceEndpointsLLM
+from distilabel.pipeline import Pipeline
+from distilabel.steps import KeepColumns
+from distilabel.steps.tasks import MagpieGenerator
+
+MODEL = "meta-llama/Meta-Llama-3.1-8B-Instruct"
+
+
+class InstructionResponsePipeline:
+    """Generates instructions and responses for a given system prompt.
+
+    This example pipeline can be used for a Supervised Fine-Tuning dataset which you
+    could use to train or evaluate a model. The pipeline generates instructions using the
+    MagpieGenerator and responses for a given system prompt. The pipeline then keeps only
+    the instruction, response, and model_name columns.
+
+    References:
+        - [Magpie: Alignment Data Synthesis from Scratch by Prompting Aligned LLMs with Nothing](https://arxiv.org/abs/2406.08464)
+
+    Example:
+
+        Generate instructions and responses for a given system prompt:
+
+        ```python
+        from distilabel.pipeline import InstructionResponsePipeline
+
+        pipeline = InstructionResponsePipeline()
+
+        distiset = pipeline.run()
+        ```
+
+        Customizing the pipeline further:
+
+        ```python
+        from distilabel.pipeline import InstructionResponsePipeline
+
+        pipeline = InstructionResponsePipeline(
+            system_prompt="You are a creative AI Assistant for writing science fiction.",
+            llm=InferenceEndpointsLLM(
+                model_id="meta-llama/Meta-Llama-3.2-3B-Instruct",
+                tokenizer_id="meta-llama/Meta-Llama-3.2-3B-Instruct",
+                generation_kwargs={"max_new_tokens": 512, "temperature": 0.7},
+            ),
+            num_rows=500,
+            batch_size=2,
+            n_turns=2,
+        )
+        ```
+    """
+
+    def __init__(
+        self,
+        llm: Optional[LLM] = None,
+        system_prompt: str = "You are a creative AI Assistant writer.",
+        hf_token: Optional[str] = None,
+        n_turns: int = 1,
+        num_rows: int = 10,
+        batch_size: int = 1,
+    ) -> None:
+        if llm is None:
+            self.llm: LLM = InferenceEndpointsLLM(
+                model_id=MODEL,
+                tokenizer_id=MODEL,
+                magpie_pre_query_template="llama3",
+                generation_kwargs={
+                    "temperature": 0.9,
+                    "do_sample": True,
+                    "max_new_tokens": 2048,
+                    "stop_sequences": [
+                        "<|eot_id|>",
+                        "<|start_header_id|>",
+                        "assistant",
+                        " \n\n",
+                    ],
+                },
+                api_key=hf_token,
+            )
+        else:
+            if isinstance(
+                llm, InferenceEndpointsLLM
+            ) and llm.magpie_pre_query_template not in [
+                "llama3",
```

Comment:
@burtenshaw I don't think this is limited to the InferenceEndpointsLLM class but rather a general pitfall for misconfiguration. 

I believe users need to set 1) the pre_query_template to either qwen2 or llama3, 2) they need to provide a tokenizer too 3) they need to ensure that the model they used is actually using the same template as the provided model/tokenizer. 

I am not sure if these 3 scenarios are handled by some error logic for distilabel internally already.

Reply:
@gabrielmbmb @plaguss could you suggest some simple logic to validate the LLM passed to the pipeline. I think we should start with something strict and simple.

Reply:
The `LLM` is already validated by `MagpieGenerator.model_post_init`, and throw an error with the part of the documentation to visit if the `LLM` is not as expected, so we can remove the `else` clause directly.